### ISSUE TITLE
memory overflow when using numpy load in a loop (migrated from Trac #2048)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#3595]
Original ticket http://projects.scipy.org/numpy/ticket/2048
Reported 2012-02-11 by trac user eldada, assigned to unknown.

Looping over npz files load causes memory overflow (depending on the file list length).
None of the following seems to help[[BR]]
1. Deleting the variable which stores the data in the file.[[BR]]
2. Using mmap.[[BR]]
3. calling gc.collect() (garbage collection).[[BR]]
   [[BR]]

The following code should reproduce the phenomenon:[[BR]]

import numpy as np[[BR]]
# generate a file for the demo[[BR]]

X = np.random.randn(1000,1000)[[BR]]

np.savez('tmp.npz',X=X)[[BR]]
[[BR]]
# here come the overflow:[[BR]]

for i in xrange(1000000):[[BR]]

```
data = np.load('tmp.npz')[[BR]]

data.close()  # avoid the "too many files are open" error[[BR]]
```

[[BR]]
# in my real application the loop is over a list of files and the overflow exceeds 24GB of RAM! [[BR]]
# please note that this was tried on ubuntu 11.10, and for both numpy v 1.5.1 as well as 1.6.0

**Comment by thouis** [comment]
Comment in Trac by atmention:pv, 2012-02-12

Adding `gc.collect()` inside the loop fixes the issue for me. There's indeed a reference cycle, but that's not a real bug, as the garbage collector catches it. (Although we could set `self.f = None` inside `close()` to break the cycle automatically.)

**Comment by thouis** [comment]
Comment in Trac by trac user eldada, 2012-02-12

 I have posted the question on http://stackoverflow.com/a/9245218/1204331 and received an answer with a solution + a general way to solve similar issues.
 The specific solution is to delete data.f 
After testing the solution proposed by pv (to gc.collect in the loop) I found that the del data.f solution is about 2000 times faster. The other proposed solution by pv (data.f=None) is as fast as deletion. I wonder why would the NpzFile class would be coded in a cyclic referring manner.
